File: day05/aoc5-1.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $INPUT_FILE = "aocdata.txt";

open my $fh, '<', $INPUT_FILE or die "Cannot open $INPUT_FILE: $!";
my @lines = ();
while (my $line = <$fh>) {
    chomp $line;
    push @lines, $line;
}
close $fh;

# Split ranges and IDs at blank line
my $split_index;
for my $i (0..$#lines) {
    if ($lines[$i] =~ /^\s*$/) {
        $split_index = $i;
        last;
    }
}
die "No blank line separating ranges and IDs" unless defined $split_index;

my @ranges = @lines[0..$split_index-1];
my @ids    = @lines[$split_index+1..$#lines];

sub parse_ranges {
    my @range_lines = @_;
    my @ranges;
    for my $line (@range_lines) {
        my ($start, $end) = split /-/, $line;
        push @ranges, [$start, $end];
    }
    return @ranges;
}

sub is_fresh {
    my ($id, $ranges_ref) = @_;
    for my $r (@$ranges_ref) {
        my ($start, $end) = @$r;
        return 1 if $id >= $start && $id <= $end;
    }
    return 0;
}

my @parsed_ranges = parse_ranges(@ranges);
my @fresh_ids = grep { is_fresh($_, \@parsed_ranges) } @ids;

print "Total available IDs: " . scalar(@ids) . "\n";
print "Fresh IDs: " . scalar(@fresh_ids) . "\n";
# print "List of fresh IDs: " . join(", ", @fresh_ids) . "\n";

File: day05/aoc5-1.py
#!/usr/bin/env python3

INPUT_FILE = "aocdata.txt"

def read_database(path):
    with open(path, 'r', encoding='utf-8') as f:
        lines = [line.strip() for line in f if line.strip() != "" or line == "\n"]
    # Split ranges and IDs at the blank line
    if "" in lines:
        split_index = lines.index("")
        ranges = lines[:split_index]
        ids = lines[split_index+1:]
    else:
        raise ValueError("No blank line separating ranges and IDs")
    return ranges, ids

def parse_ranges(range_lines):
    ranges = []
    for line in range_lines:
        start, end = map(int, line.split('-'))
        ranges.append((start, end))
    return ranges

def parse_ids(id_lines):
    return [int(x) for x in id_lines]

def is_fresh(id_val, ranges):
    for start, end in ranges:
        if start <= id_val <= end:
            return True
    return False

def main():
    ranges_raw, ids_raw = read_database(INPUT_FILE)
    ranges = parse_ranges(ranges_raw)
    ids = parse_ids(ids_raw)

    fresh_ids = [id_val for id_val in ids if is_fresh(id_val, ranges)]
    print(f"Total available IDs: {len(ids)}")
    print(f"Fresh IDs: {len(fresh_ids)}")
#    print(f"List of fresh IDs: {fresh_ids}")

if __name__ == "__main__":
    main()
